exporting/mongodb/mongodb.c

```diff
@@ -0,0 +1,310 @@
+        error("EXPORTING: failed to create a new client");
+        return 1;
+    }
+
+    if (!mongoc_client_set_appname(connector_specific_data->client, "netdata")) {
+        error("EXPORTING: failed to set client appname");
+    };
+
+    connector_specific_data->collection = mongoc_client_get_collection(
+        connector_specific_data->client, connector_specific_config->database, connector_specific_config->collection);
+
+    mongoc_uri_destroy(uri);
+
+    // create a ring buffer
+    struct bson_buffer *first_buffer = NULL;
+
+    if (instance->config.buffer_on_failures < 2)
+        instance->config.buffer_on_failures = 1;
+    else
+        instance->config.buffer_on_failures -= 1;
+
+    for (int i = 0; i < instance->config.buffer_on_failures; i++) {
+        struct bson_buffer *current_buffer = callocz(1, sizeof(struct bson_buffer));
+
+        if (!connector_specific_data->first_buffer)
+            first_buffer = current_buffer;
+        else
+            current_buffer->next = connector_specific_data->first_buffer;
+
+        connector_specific_data->first_buffer = current_buffer;
+    }
+
+    first_buffer->next = connector_specific_data->first_buffer;
+    connector_specific_data->last_buffer = connector_specific_data->first_buffer;
+
+    return 0;
+}
+
+/**
+ * Clean a MongoDB connector instance up
+ *
+ * @param instance an instance data structure.
+ */
+void mongodb_cleanup(struct instance *instance)
+{
+    struct mongodb_specific_data *connector_specific_data =
+        (struct mongodb_specific_data *)instance->connector_specific_data;
+
+    mongoc_collection_destroy(connector_specific_data->collection);
+    mongoc_client_destroy(connector_specific_data->client);
+
+    return;
+}
+
+/**
+ * Initialize a MongoDB connector instance
+ *
+ * @param instance an instance data structure.
+ * @return Returns 0 on success, 1 on failure.
+ */
+int init_mongodb_instance(struct instance *instance)
+{
+    instance->worker = mongodb_connector_worker;
+
+    instance->start_batch_formatting = NULL;
+    instance->start_host_formatting = format_host_labels_json_plaintext;
+    instance->start_chart_formatting = NULL;
+
+    if (EXPORTING_OPTIONS_DATA_SOURCE(instance->config.options) == EXPORTING_SOURCE_DATA_AS_COLLECTED)
+        instance->metric_formatting = format_dimension_collected_json_plaintext;
+    else
+        instance->metric_formatting = format_dimension_stored_json_plaintext;
+
+    instance->end_chart_formatting = NULL;
+    instance->end_host_formatting = flush_host_labels;
+    instance->end_batch_formatting = format_batch_mongodb;
+
+    instance->send_header = NULL;
+    instance->check_response = NULL;
+
+    instance->buffer = (void *)buffer_create(0);
+    if (!instance->buffer) {
+        error("EXPORTING: cannot create buffer for MongoDB exporting connector instance %s", instance->config.name);
+        return 1;
+    }
+    uv_mutex_init(&instance->mutex);
+    uv_cond_init(&instance->cond_var);
```

Comment:
Both `uv_mutex_init` and `uv_cond_init` returns integer, should we add a test for the return here?

Reply:
Fixed